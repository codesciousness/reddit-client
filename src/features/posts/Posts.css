.Posts__trending ul {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-wrap: wrap;
}

.Posts__trending__error {
    margin: 2rem 0;
}

.Posts__trending li {
    list-style: none;
}

.Posts__latest {
    width: 60%;
    margin: 4rem auto auto auto;
}

.Posts__latest__error {
    margin: 2rem 0;
}

.Posts__latest li {
    list-style: none;
}