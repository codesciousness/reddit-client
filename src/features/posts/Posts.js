import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import './Posts.css';
import Post from '../../components/post/Post';
import Loader from '../../components/loader/Loader';
import { selectSearchTerm } from '../searchTerm/searchTermSlice';
import { selectFilter } from '../filter/filterSlice';
import { selectFilteredTrendingPosts,
        selectFilteredLatestPosts,
        selectLoadingTrendingPosts,
        selectLoadTrendingPostsError,
        selectLoadingLatestPosts,
        selectLoadLatestPostsError,
        loadTrendingPosts,
        loadLatestPosts } from './postsSlice';

const Posts = ({ isTrending }) => {
    const searchTerm = useSelector(selectSearchTerm);
    const filter = useSelector(selectFilter);
    const trendingPosts = useSelector(selectFilteredTrendingPosts);
    const latestPosts = useSelector(selectFilteredLatestPosts);
    const loadingTrendingPosts = useSelector(selectLoadingTrendingPosts);
    const loadingLatestPosts = useSelector(selectLoadingLatestPosts);
    const loadTrendingPostsError = useSelector(selectLoadTrendingPostsError);
    const loadLatestPostsError = useSelector(selectLoadLatestPostsError);
    const dispatch = useDispatch();

    useEffect(() => {
        dispatch(loadTrendingPosts());
        dispatch(loadLatestPosts());
    }, [filter, searchTerm, dispatch]);

    if (isTrending) {
        if (loadingTrendingPosts) {
            return (
                <section className="Posts__trending">
                    <Loader />
                </section>
            );
        }
        if (loadTrendingPostsError) {
            return (
                <section className="Posts__trending">
                    <p className="Posts__trending__error">Request failed! Please reload.</p>
                </section>
            );
        }
        return (
            <section className="Posts__trending">
                {trendingPosts.map(post => <Post post={post} isTrending={true} />)}
            </section>
        );
    }
    else {
        if (loadingLatestPosts) {
            return (
                <section className="Posts__latest">
                    <Loader />
                </section> 
            );
        }
        if (loadLatestPostsError) {
            return (
                <section className="Posts__latest">
                    <p className="Posts__latest__error">Request failed! Please reload.</p>
                </section> 
            );
        }
        return (
            <section className="Posts__latest">
                {latestPosts.map(post => <Post post={post} isTrending={false} />)}
            </section> 
        );
    }
}

export default Posts;