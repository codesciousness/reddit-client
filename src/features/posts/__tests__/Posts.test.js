import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Posts from '../Posts';

const mockStore = configureStore([thunk]);

const trendingPosts = [
    {
        id: 'efgh2i',
        title: 'Where Can the Solar Eclipse be Viewed Next Week',
        author: 'ScienceResearcher',
        created: '4 hours ago',
        flair: 'Astronomy',
        thumbnail: 'https://source.unsplash.com/140x140/?astronomy',
        commentNum: 25,
        url: 'https://www.sciencemag.org/'
    },
    {
        id: 'jklm3n',
        title: 'An Overview of Traditional Diets versus Modern Diets',
        author: 'ScienceLover',
        created: '2 hours ago',
        flair: 'Health',
        thumbnail: 'https://source.unsplash.com/140x140/?health',
        commentNum: 6,
        url: 'https://www.nutrition.org/'
    }
];

const latestPosts = [
    {
        id: 'opqr4s',
        title: 'Causes of Reproductive Stress on Animals in Declining Habitats',
        author: 'ScienceGuy',
        created: '25 minutes ago',
        flair: 'Biology',
        thumbnail: 'https://source.unsplash.com/140x140/?biology',
        commentNum: 2,
        url: 'https://www.nature.com/'
    },
    {
        id: 'tuvw5x',
        title: 'Activation of the Brain in Subjects With High And Average Intelligence',
        author: 'ScienceObserver',
        created: '40 minutes ago',
        flair: 'Neuroscience',
        thumbnail: 'https://source.unsplash.com/140x140/?neuroscience',
        commentNum: 3,
        url: 'https://www.jneurosci.org/'
    }
];

describe('<Posts />', () => {
    let store;

    beforeEach(() => {
        store = mockStore({
            filter: '',
            searchTerm: '',
            posts: {
                trendingPosts,
                latestPosts,
                loadingTrendingPosts: false,
                loadTrendingPostsError: false,
                loadingLatestPosts: false,
                loadLatestPostsError: false,
            }
        });
    });

    test('Posts component is defined', () => {
        expect(<Posts />).toBeDefined();
    });
    test('matches the snapshot when isTrending=true prop is passed', () => {
        const component = create(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('matches the snapshot when isTrending=false prop is passed', () => {
        const component = create(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={false} />
                </Router>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly when isTrending=true prop is passed', () => {
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.Posts__trending')).toBeInTheDocument();
    });
    test('renders correctly when isTrending=false prop is passed', () => {
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={false} />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.Posts__latest')).toBeInTheDocument();
    });
    test('renders and passes each trending post from state as a prop to a Post component', () => {
        render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        expect(screen.getByText(/Where Can the Solar Eclipse be Viewed/)).toBeInTheDocument();
        expect(screen.getByText(/An Overview of Traditional Diets versus Modern/)).toBeInTheDocument();
    });
    test('renders and passes each latest post from state as a prop to a Post component', () => {
        render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={false} />
                </Router>
            </Provider>
        );
        expect(screen.getByText(/Causes of Reproductive Stress on Animals in Declining Habitats/)).toBeInTheDocument();
        expect(screen.getByText(/Activation of the Brain in Subjects With High And Average Intelligence/)).toBeInTheDocument();
    });
    test('dispatches async action creators on initial render', () => {
        render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        const actions = store.getActions();
        const firstActionType = actions[0].type;
        const secondActionType = actions[1].type;
        const dispatched = [ firstActionType, secondActionType];
        const expected = ['posts/loadTrendingPosts/pending', 'posts/loadLatestPosts/pending'];
        expect(dispatched).toEqual(expected);
    });
    test('dispatches async action creators when filter changes', () => {
        render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        let actions = store.getActions();
        expect(actions.length).toBe(2);
        store.filter = 'Health';
        render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        actions = store.getActions();
        expect(actions.length).toBe(4);
        const firstActionType = actions[0].type;
        const secondActionType = actions[1].type;
        const thirdActionType = actions[2].type;
        const fourthActionType = actions[3].type;
        const dispatched = [ firstActionType, secondActionType, thirdActionType, fourthActionType];
        const expected = ['posts/loadTrendingPosts/pending', 'posts/loadLatestPosts/pending', 'posts/loadTrendingPosts/pending', 'posts/loadLatestPosts/pending'];
        expect(dispatched).toEqual(expected);
    });
    test('dispatches async action creators when searchTerm changes', () => {
        render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        let actions = store.getActions();
        expect(actions.length).toBe(2);
        store.searchTerm = 'c';
        render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        actions = store.getActions();
        expect(actions.length).toBe(4);
        const firstActionType = actions[0].type;
        const secondActionType = actions[1].type;
        const thirdActionType = actions[2].type;
        const fourthActionType = actions[3].type;
        const dispatched = [ firstActionType, secondActionType, thirdActionType, fourthActionType];
        const expected = ['posts/loadTrendingPosts/pending', 'posts/loadLatestPosts/pending', 'posts/loadTrendingPosts/pending', 'posts/loadLatestPosts/pending'];
        expect(dispatched).toEqual(expected);
    });
    test('renders a Loader component under trending posts when loadingTrendingPosts=true', () => {
        store = mockStore({
            filter: '',
            searchTerm: '',
            posts: {
                trendingPosts,
                latestPosts,
                loadingTrendingPosts: true,
                loadTrendingPostsError: false,
                loadingLatestPosts: false,
                loadLatestPostsError: false,
            }
        });
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.Posts__trending')).toBeInTheDocument();
        expect(container.querySelector('.Loader')).toBeInTheDocument();
    });
    test('renders a Loader component under latest posts when loadingLatestPosts=true', () => {
        store = mockStore({
            filter: '',
            searchTerm: '',
            posts: {
                trendingPosts,
                latestPosts,
                loadingTrendingPosts: false,
                loadTrendingPostsError: false,
                loadingLatestPosts: true,
                loadLatestPostsError: false,
            }
        });
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={false} />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.Posts__latest')).toBeInTheDocument();
        expect(container.querySelector('.Loader')).toBeInTheDocument();
    });
    test('renders error message under trending posts when loadTrendingPostsError=true', () => {
        store = mockStore({
            filter: '',
            searchTerm: '',
            posts: {
                trendingPosts,
                latestPosts,
                loadingTrendingPosts: false,
                loadTrendingPostsError: true,
                loadingLatestPosts: false,
                loadLatestPostsError: false,
            }
        });
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={true} />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.Posts__trending')).toBeInTheDocument();
        expect(container.querySelector('.Posts__trending__error')).toBeInTheDocument();
        expect(screen.getByText(/Error: Request failed! Please try again./)).toBeInTheDocument();
    });
    test('renders error message under latest posts when loadLatestPostsError=true', () => {
        store = mockStore({
            filter: '',
            searchTerm: '',
            posts: {
                trendingPosts,
                latestPosts,
                loadingTrendingPosts: false,
                loadTrendingPostsError: false,
                loadingLatestPosts: false,
                loadLatestPostsError: true,
            }
        });
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Posts isTrending={false} />
                </Router>
            </Provider>
        );
        expect(container.querySelector('.Posts__latest')).toBeInTheDocument();
        expect(container.querySelector('.Posts__latest__error')).toBeInTheDocument();
        expect(screen.getByText(/Error: Request failed! Please try again./)).toBeInTheDocument();
    });
});