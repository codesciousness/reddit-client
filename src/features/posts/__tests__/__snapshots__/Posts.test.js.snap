// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts /> matches the snapshot when isTrending=false prop is passed 1`] = `
<section
  className="Posts__latest"
>
  <ul>
    <li>
      <section
        className="Post__latest"
      >
        <img
          alt=""
          className="Post__latest__image"
          src="https://source.unsplash.com/140x140/?biology&sig=opqr4s"
        />
        <div
          className="Post__latest__content"
        >
          <p
            className="Post__latest__flair"
          >
            Biology
          </p>
          <a
            className="Post__latest__link"
            href="/posts/biology/opqr4s"
            onClick={[Function]}
          >
            <h2
              className="Post__latest__title"
              id="opqr4s"
              onClick={[Function]}
            >
              Causes of Reproductive Stress on Animals in Declining Habitats
            </h2>
          </a>
          <a
            href="https://www.nature.com/"
            rel="noreferrer"
            target="_blank"
          >
            <button
              className="Post__latest__button"
            >
              Read more 
              <i
                className="fas fa-angle-double-right Post__latest__icon"
              />
            </button>
          </a>
          <p
            className="Post__latest__detail"
          >
            <span
              className="Post__latest__author"
            >
              Posted by 
              <strong>
                ScienceGuy
              </strong>
            </span>
            <span
              className="PostTime"
            >
              NaN years ago
            </span>
            <span
              className="Post__latest__comments"
            >
              <i
                className="far fa-comment-alt"
              />
               
              2
            </span>
          </p>
        </div>
      </section>
    </li>
    <li>
      <section
        className="Post__latest"
      >
        <img
          alt=""
          className="Post__latest__image"
          src="https://source.unsplash.com/140x140/?neuroscience&sig=tuvw5x"
        />
        <div
          className="Post__latest__content"
        >
          <p
            className="Post__latest__flair"
          >
            Neuroscience
          </p>
          <a
            className="Post__latest__link"
            href="/posts/neuroscience/tuvw5x"
            onClick={[Function]}
          >
            <h2
              className="Post__latest__title"
              id="tuvw5x"
              onClick={[Function]}
            >
              Activation of the Brain in Subjects With High And Average Intelligence
            </h2>
          </a>
          <a
            href="https://www.jneurosci.org/"
            rel="noreferrer"
            target="_blank"
          >
            <button
              className="Post__latest__button"
            >
              Read more 
              <i
                className="fas fa-angle-double-right Post__latest__icon"
              />
            </button>
          </a>
          <p
            className="Post__latest__detail"
          >
            <span
              className="Post__latest__author"
            >
              Posted by 
              <strong>
                ScienceObserver
              </strong>
            </span>
            <span
              className="PostTime"
            >
              NaN years ago
            </span>
            <span
              className="Post__latest__comments"
            >
              <i
                className="far fa-comment-alt"
              />
               
              3
            </span>
          </p>
        </div>
      </section>
    </li>
  </ul>
</section>
`;

exports[`<Posts /> matches the snapshot when isTrending=true prop is passed 1`] = `
<section
  className="Posts__trending"
>
  <ul>
    <li>
      <section
        className="Post__trending"
      >
        <img
          alt=""
          className="Post__trending__image"
          src="https://source.unsplash.com/140x140/?astronomy&sig=efgh2i"
        />
        <div
          className="Post__trending__content"
        >
          <h2
            className="Post__trending__flair"
          >
            Astronomy
          </h2>
          <a
            className="Post__trending__link"
            href="/posts/astronomy/efgh2i"
            onClick={[Function]}
          >
            <h3
              className="Post__trending__title"
              id="efgh2i"
              onClick={[Function]}
            >
              Where Can the Solar Eclipse be Viewed...
            </h3>
          </a>
        </div>
      </section>
    </li>
    <li>
      <section
        className="Post__trending"
      >
        <img
          alt=""
          className="Post__trending__image"
          src="https://source.unsplash.com/140x140/?health&sig=jklm3n"
        />
        <div
          className="Post__trending__content"
        >
          <h2
            className="Post__trending__flair"
          >
            Health
          </h2>
          <a
            className="Post__trending__link"
            href="/posts/health/jklm3n"
            onClick={[Function]}
          >
            <h3
              className="Post__trending__title"
              id="jklm3n"
              onClick={[Function]}
            >
              An Overview of Traditional Diets versus Modern...
            </h3>
          </a>
        </div>
      </section>
    </li>
  </ul>
</section>
`;
